common/types.go

```diff
@@ -373,11 +373,37 @@ func (a Address) Value() (driver.Value, error) {
 	return a[:], nil
 }
 
-// Checks if an address is a valid account in our node's sharded address space
+// IsInChainScope checks if an address is a valid account in our node's sharded address space
 func (a Address) IsInChainScope() bool {
 	return NodeLocation.ContainsAddress(a)
 }
 
+// Location looks up the chain location which contains this address
+func (a Address) Location() *Location {
```

Comment:
shouldn't this be a switch case since we know when to call which of these logical blocks?.

Reply:
I'm not sure I understand. Given an address, we want to see what location this address belongs to. We don't know anything about the address, not even which context it is.

Reply:
Why can't we create a byte array similar to what we had before that we can index an address in to?

Reply:
a bit of an aggressive comment but like so:

```
// bytePrefixList expands the space with proper size for chain IDs to be indexed
var (
	mainnetBytePrefixList = make([][]int, 20000)
	testnetBytePrefixList = make([][]int, 20000)
)

func init() {
	mainnetBytePrefixList[9000] = []int{0, 9}
	mainnetBytePrefixList[9100] = []int{10, 19}
	mainnetBytePrefixList[9101] = []int{20, 29}
	mainnetBytePrefixList[9102] = []int{30, 39}
	mainnetBytePrefixList[9103] = []int{40, 49}
	mainnetBytePrefixList[9200] = []int{50, 59}
	mainnetBytePrefixList[9201] = []int{60, 69}
	mainnetBytePrefixList[9202] = []int{70, 79}
	mainnetBytePrefixList[9203] = []int{80, 89}
	mainnetBytePrefixList[9300] = []int{90, 99}
	mainnetBytePrefixList[9301] = []int{100, 109}
	mainnetBytePrefixList[9302] = []int{110, 119}
	mainnetBytePrefixList[9303] = []int{120, 129}

	testnetBytePrefixList[12000] = []int{0, 9}
	testnetBytePrefixList[12100] = []int{10, 19}
	testnetBytePrefixList[12101] = []int{20, 29}
	testnetBytePrefixList[12102] = []int{30, 39}
	testnetBytePrefixList[12103] = []int{40, 49}
	testnetBytePrefixList[12200] = []int{50, 59}
	testnetBytePrefixList[12201] = []int{60, 69}
	testnetBytePrefixList[12202] = []int{70, 79}
	testnetBytePrefixList[12203] = []int{80, 89}
	testnetBytePrefixList[12300] = []int{90, 99}
	testnetBytePrefixList[12301] = []int{100, 109}
	testnetBytePrefixList[12302] = []int{110, 119}
	testnetBytePrefixList[12303] = []int{120, 129}
}

// LookupChainIDRange returns the byte lookup based off a configs chainID
func LookupChainIDRange(index *big.Int) []int {
	for _, inSet := range mainnetValidChains {
		if index.Cmp(inSet) == 0 {
			lookup := mainnetBytePrefixList[index.Int64()]
			return lookup
		}
	}

	for _, inSet := range testnetValidChains {
		if index.Cmp(inSet) == 0 {
			lookup := testnetBytePrefixList[index.Int64()]
			return lookup
		}
	}
	return nil
}
```

Reply:
> Why can't we create a byte array map similar to what we had before that we can index an address in to?

We do have such a map, here: https://github.com/dominant-strategies/go-quai/blob/d64fa1534aacc91a36b16680e7063214edfcf9c7/common/types.go#L218

And this prefix map is what is used under the hood in this function when calling `ContainsAddress(...)`. The intended usecase for this method is to map an address to a location, i.e:
```
address  := ...                // given some address
location := address.Location() // lookup the chain that address belongs to
```
I'm still not clear what your proposal is. Namely, how do you propose to use `LookupChainIDRange(index big.Int) []int` to map an `Address` into a `Location` (or chain ID)? One observation is that my approach only searches the address space once, but yours has to search many times for mainnet IDs, testnet IDs (and more testnet/devnet IDs in the future?).

There may be better approaches to the mapping, but this was simultaneously the most readable and most performant approach I could think of. I may just not understanding your suggestion completely.

Reply:
One other thing in the back of my mind, is that this structure will be very convenient when we redo our address mapping, [which needs to happen before launch](https://github.com/dominant-strategies/quai-specs/issues/24).

Reply:
Since we cannot key off of the []byte. The current approach is better.